include/net/nrf_cloud_logs.h

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef NRF_CLOUD_LOGS_H_
+#define NRF_CLOUD_LOGS_H_
+
+/** @file nrf_cloud_logs.h
+ * @brief Module to provide nRF Cloud logging support to nRF9160 SiP.
+ */
+
+#include <zephyr/logging/log.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct nrf_cloud_rest_context;
+
+/** @defgroup nrf_cloud_logs nRF Cloud Logs
+ * @{
+ */
+
+/** @brief Initialize the nRF Cloud logging subsystem. Call after
+ *  the date_time library is ready.
+ */
+void nrf_cloud_log_init(void);
+
+/** @brief Set log_level to the level at or below which a log message will be
+ *  sent to the cloud. Log messages with a log level larger than this
+ *  will not be sent. For example, if log_level is LOG_LEVEL_WRN (2), then
+ *  LOG_LEVEL_WRN (2) and LOG_LEVEL_ERR (1) messages will be sent, and
+ *  LOG_LEVEL_INF (3) and LOG_LEVEL_DBG (4) will not.
+ *  Set log_level to LOG_LEVEL_NONE (0) to disable logging.
+ *  See zephyr/include/zephyr/logging/log_core.h.
+ *
+ * @param log_level The log level to go to the cloud.
+ */
+void nrf_cloud_log_control_set(int log_level);
+
+/** @brief Get current log level to be sent to nRF Cloud.
+ * @retval Current log level, if CONFIG_NRF_CLOUD_LOG_BACKEND is enabled, otherwise
+ *         0, indicating logs are disabled.
+ */
+int nrf_cloud_log_control_get(void);
+
+/** @brief Tell REST-based logger the REST context and device_id for later
+ *         use when outputting a log.
+ *
+ *  @param[in] ctx REST context to use.
+ *  @param[in] dev_id Device ID to send message on behalf of.
+ */
+void nrf_cloud_rest_log_context_set(struct nrf_cloud_rest_context *ctx, const char *dev_id);
+
+/** @brief Enable or disable logging to the cloud.
+ *  @param[in] enable Set true to send logs to the cloud, false to disable.
+ */
+void nrf_cloud_log_enable(bool enable);
+
+/** @brief Find out if logger is enabled.
+ *
+ *  @retval True if logger is enabled.
+ */
+bool nrf_cloud_log_is_enabled(void);
+
+#if defined(CONFIG_NRF_CLOUD_MQTT)
+/** @brief Directly log to the cloud. This does not use the Zephyr logging
+ *  system if CONFIG_NRF_CLOUD_LOG_BACKEND is disabled.  Otherwise, it is passed to
+ *  the logging system.
```

Comment:
See https://github.com/nrfconnect/sdk-nrf/pull/10266#pullrequestreview-1322686250.

Reply:
OK, this is a bit confusing when looking at the API at first but I guess documentation will make it clearer.

> i would request the ability for a user to use logging without tying into the logger backend.
> `nrf_cloud_logs_send(DBG, "something something happened: %d", ret_val);`
> `nrf_cloud_rest_logs_send(&rest_ctx, DBG, "something something happened: %d", ret_val);`
> i feel like this would be useful for actual log messages the user wants to track from a device, instead of for debugging.

I'm still thinking why wouldn't application developer do the same with Zephyr logger by using a separate logger instance for "actual log messages the user wants to track from a device". Am I missing something that this brings on top of that?

Reply:
I think the reason @jayteemo wanted this is his concern that using the logging backend risks customers inadvertently using a lot of data due to misconfigured logging levels for unimportant subsystems.  By offering this direct-logging-only mechanism, the customer can still benefit from our portal's front end for viewing logs, but can be assured that truly only inforrmation they care about is sent there.

Reply:
Forgot to comment this yesterday.
If I'm not missing something, doing the "direct log messages" without additional logging from other modules would require the following configuration (which we could guide in the documentation):
CONFIG_LOG_DEFAULT_LEVEL=0
CONFIG_<APPNAME>_LOG_LEVEL_DBG=y

If it's as simple as this, I'm not necessarily seeing the benefits of creating own functions for doing the same thing.
Here is my list of drawbacks of not using existing logging implementation for this:

- Two ways of doing the same thing.
- Testing required for 3 different configurations (both enabled, or one of them enabled)
- Some level of risk of having to fix some bug(s) in `CONFIG_NRF_CLOUD_LOG_DIRECT` implementation.
- A bit more complicated configuration and implementation within nrf_cloud_log, which makes it slightly less understandable also for the user when (s)he reads the document and function/Kconfig comments (I'm not saying it's very complicated but it's more complicated than without this feature).

It's your choice but take your time and think about the pros and cons.

Whether or not `CONFIG_NRF_CLOUD_LOG_DIRECT` implementation will remain there, I'd add a note to the documentation that `CONFIG_LOG_DEFAULT_LEVEL` can be used to control the default logging level sent to the cloud by all modules whose logging level is not specifically set when using `CONFIG_NRF_CLOUD_LOG_BACKEND`.

Reply:
I think you do not understand the purpose or how these features work.

`CONFIG_NRF_CLOUD_LOG_BACKEND=n` results in the nRF Cloud logging backend being not compiled into the project. There will be no Zephyr log messages sent to nRF Cloud in this configuration (no LOG_XXX() messages). There is no need to set `CONFIG_LOG_DEFAULT_LEVEL=0`.  This has nothing to do with disabling all of Zephyr logging for all logging backends -- only whether the cloud logging backend is available or not.

`CONFIG_NRF_CLOUD_LOG_DIRECT=y` and `CONFIG_NRF_CLOUD_LOG_BACKEND=n`  enables the direct function calls to send the same format of JSON log messages to nRF Cloud as are used when the nRF Cloud logging backend is enabled.  They are sent using `nrf_cloud_send()` or `nrf_cloud_rest_send_device_message()` directly.  The Zephyr logging system is unused for these messages. This results in the portal working as expected, and allows the customer to intentionally send log messages to the cloud purposefully, not implicitly by the logging system.

If both `CONFIG_NRF_CLOUD_LOG_BACKEND=y` and `CONFIG_NRF_CLOUD_LOG_DIRECT=y`, then for convenience, I route the messages sent to the direct function calls into the logging system so they do not appear out of time order compared to LOG_XXX() messages, as they would if the direct call was done immediately. It also lets a customer include the direct calls and not have to put conditional compilation around them if they sometimes do a build that does use the full logging option.

Yes, this results in two ways of doing the same thing, but with different side effects, power use, and cellular data use, so there are benefits to providing both.  Yes, it increases testing.  Yes, it's a bit more complicated.

We have though about the pros and cons and are comfortable with maintaining this system.  It is in the nRF Cloud library so it is our responsibility as code owners to do so.

I have manually tested this feature in all 3 configurations and all 3 work as I planned.

Reply:
> I think you do not understand the purpose or how these features work.

I have understood the features as you wrote. I didn't when I started my first review but I've understood them for a while now.

> `CONFIG_NRF_CLOUD_LOG_BACKEND=n` results in the nRF Cloud logging backend being not compiled into the project. There will be no Zephyr log messages sent to nRF Cloud in this configuration (no LOG_XXX() messages). There is no need to set `CONFIG_LOG_DEFAULT_LEVEL=0`. This has nothing to do with disabling all of Zephyr logging for all logging backends -- only whether the cloud logging backend is available or not.

Yep, `CONFIG_LOG_DEFAULT_LEVEL=0` is not relevant when `CONFIG_NRF_CLOUD_LOG_BACKEND=n`. I was just saying that you can achieve the same functionality of `CONFIG_NRF_CLOUD_LOG_DIRECT=y` + `CONFIG_NRF_CLOUD_LOG_BACKEND=n` with `CONFIG_NRF_CLOUD_LOG_DIRECT=n` + `CONFIG_NRF_CLOUD_LOG_BACKEND=y` + `CONFIG_LOG_DEFAULT_LEVEL=0` + `CONFIG_APPNAME_LOG_LEVEL_DBG=y`.

And I mentioned that there could be a note in the documentation to advise users to think about using `CONFIG_LOG_DEFAULT_LEVEL` when `CONFIG_NRF_CLOUD_LOG_BACKEND=y`.

> Yes, this results in two ways of doing the same thing, but with different side effects, power use, and cellular data use, so there are benefits to providing both.

I assume that with power use you refer to convenience of using `nrf_cloud_send()` or `nrf_cloud_rest_send_device_message()` instead of `LOG_XXX()`. Or is it something else?

How is cellular data use better in one configuration over the other assuming you log the same information?
Buffering done by Zephyr logging system gives some benefits on the use of data link when `CONFIG_NRF_CLOUD_LOG_BACKEND=y`.

Reply:
I am not sure how `CONFIG_NRF_CLOUD_LOG_DIRECT=n` + `CONFIG_NRF_CLOUD_LOG_BACKEND=y` + `CONFIG_LOG_DEFAULT_LEVEL=0` + `CONFIG_APPNAME_LOG_LEVEL_DBG=y` is in any way equivalent to `CONFIG_NRF_CLOUD_LOG_DIRECT=y` + `CONFIG_NRF_CLOUD_LOG_BACKEND=n` .  If the user has specified a specific log level for one or more subsystems, then `CONFIG_LOG_DEFAULT_LEVEL=0` is ineffective at turning logs from that subsystem off.

I think customers will want to allow UART or RTT logging for various subsystems to remain active, but disable logging to nRF Cloud, and only intentionally log specific messages

Reply:
From the application logging perspective they are equivalent.
True, if you want more UART logging from subsystems and only your application logging to the cloud, you'll need the "direct messages". I guess you cannot change from the cloud between only application logging and application+subsystem logging at runtime in such configuration.
And you cannot get a specific subsystem logging to cloud, e.g., if I'd be interested in Location library logging but nothing else.

I guess we just have different view of the importance of this use case.

Ultimate solution would be to allow configuring logging on per subsystem basis but I understand this is more complicated in the cloud side.

Reply:
Yes, I do think there could be a mechanism defined where Zephyr runtime log filtering could be controlled from the cloud, but not for this release.